### ISSUE TITLE
[RFC] Support DiffusionNFT for BOOGU-Image and BOOGU-Image-Edit (T2I & TI2I)

### ISSUE THREAD

**Issue by SamitHuang** [verl-project/verl-omni#551]
## 1. Summary

This RFC proposes adding **DiffusionNFT** (direct-preference forward-process RL) support for the **BOOGU-Image** model family, covering both text-to-image (`Boogu-Image-0.1-Base`) and instruction-guided image editing (`Boogu-Image-0.1-Edit`).

Currently, `BooguImagePipeline` only supports policy-gradient FlowGRPO (`algorithm="flow_grpo"`). While FlowGRPO performs well, it requires collecting reverse-SDE trajectories and log-probabilities across all denoising steps, incurring high rollout host memory and network transfer overhead. DiffusionNFT trains from final clean latents ($x_0$) using a forward-process objective, enabling:
1. **Significant rollout memory reduction**: Omits trajectory latent/logprob caching, slashing rollout-to-actor tensor payload by ~10–20×.
2. **Native deterministic ODE generation**: Rollout uses native Euler/ODE inference without stochastic SDE perturbation, avoiding likelihood drift.
3. **High sample efficiency for image editing**: Provides stable preference optimization for both text-to-image synthesis and precise reference-conditioned editing.

---

### 2.1 The BOOGU-Image Model Family
BOOGU-Image is a dual-stream DiT multimodal model built on top of the Qwen3-VL text/multimodal encoder. It natively shares the `BooguImagePipeline` architecture across two primary tasks:
- **T2I (`Boogu-Image-0.1-Base`)**: Text prompts encoded via Qwen3-VL into instruction hidden states.
- **Edit / TI2I (`Boogu-Image-0.1-Edit`)**: Reference images are encoded via VAE into `condition_image_latents` and passed to the transformer as `ref_image_hidden_states`, while prompt image placeholders condition the Qwen3-VL MLLM vision tower.

### 2.2 Why DiffusionNFT for BOOGU-Image?
`verl-omni` supports DiffusionNFT on `QwenImagePipeline` and `MiniMaxH3Pipeline` via `NFTDiffusersFSDPEngine` and `DirectPreferenceRayTrainer`. Adding BOOGU-Image to this roster yields several key advantages:
- **Zero Trainer / Engine Core Changes**: `verl-omni` already provides a unified direct-preference trainer loop and FSDP2 engine (`NFTDiffusersFSDPEngine`). Only model and rollout adapters are needed.
- **T2I Request Batching**: Without SDE window constraints, T2I rollout requests can be served packed with high throughput (`supports_request_batch = True`).
- **Stable Alignment for Image Editing**: Image editing often suffers from high variance under reverse-trajectory policy gradients when the reward evaluates subtle local modifications. DiffusionNFT's clean latent reconstruction loss provides direct gradient signals anchored to the source condition.

---

## 3. Mathematical Bridge & Technical Challenges

Integrating BOOGU-Image into DiffusionNFT requires bridging three fundamental convention differences:

### 3.1 Velocity Target & Loss Inversion (Critical Pitfall)
In `NFTDiffusersFSDPEngine` and `DiffusionNFTLoss`:
1. Clean latents $x_0$ are perturbed along forward timesteps $t \in [0, 1]$ (where $t = 0$ is clean data $x_0$, and $t = 1$ is pure noise $\epsilon$):
   $$x_t = (1 - t) x_0 + t \epsilon$$
2. `DiffusionNFTLoss` reconstructs the clean latent from the predicted velocity $v$:
   $$x_{0, \text{pred}} = x_t - t \cdot v_{\text{pred}}$$
   For standard Diffusers flow matching, the velocity target is $v_{\text{diffusers}} = \epsilon - x_0$. Under this convention:
   $$x_t - t \cdot (\epsilon - x_0) = (1 - t) x_0 + t \epsilon - t \epsilon + t x_0 = x_0$$
3. **The BOOGU Inversion**: Upstream BOOGU-Image defines velocity in the opposite direction:
   $$v_{\text{boogu}} = x_0 - \epsilon = - v_{\text{diffusers}}$$
   If raw $v_{\text{boogu}}$ were passed to `DiffusionNFTLoss`, the reconstruction would evaluate to:
   $$x_t - t (x_0 - \epsilon) = (1 - 2t) x_0 + 2t \epsilon \neq x_0$$
   This would cause loss explosion and complete training divergence.
   
   **Solution**: The training adapter `forward(...)` must negate the model's raw prediction after applying text CFG:
   $$v_{\text{pred}} = - v_{\text{boogu}}$$

### 3.2 Timestep Mapping
- Diffusers/FSDPEngine timestep parameter $t \in [0, 1]$ represents noise fraction ($\sigma$).
- BOOGU-Image transformer consumes native time $t_{\text{boogu}} = 1 - \sigma$, where $0$ is noise and $1$ is clean data.
- **Solution**: The training adapter must map timesteps via `boogu_timestep_from_scheduler(timestep, num_train_timesteps)` before passing them to `BooguImageTransformer2DModel`.

### 3.3 Text CFG Formulation
BOOGU-Image uses sequential text CFG (`apply_boogu_text_cfg`) with a default guidance scale of 4.0:
$$\text{pred} = \text{negative\_pred} + \text{guidance\_scale} \times (\text{positive\_pred} - \text{negative\_pred})$$
The adapter's forward pass must execute positive and negative branches sequentially (or batched) when `pipeline.guidance_scale > 1.0`, apply `apply_boogu_text_cfg`, and then negate the result.

### 3.4 Condition Latents for Boogu-Image-Edit
For edit requests:
- Rollout generates $x_0$ conditioned on reference images and packages `condition_image_latents` (shape `(B, C, H, W)`) into the rollout RL metadata dictionary.
- During actor training, `NFTDiffusersFSDPEngine` perturbs only $x_0$ into $x_t$, keeping `condition_image_latents` unperturbed.
- The training adapter unwraps `condition_image_latents` into `ref_image_hidden_states = [[lat] for lat in condition_image_latents]` for forward, old-policy, and ref-policy passes.

---

### 4.1 New Pipeline Package: `verl_omni/pipelines/boogu_image_diffusion_nft`
Mirroring `verl_omni/pipelines/qwen_image_diffusion_nft`, the implementation will consist of:

```
verl_omni/pipelines/boogu_image_diffusion_nft/
├── __init__.py
├── diffusers_training_adapter.py  # BooguImageDiffusionNFT
└── vllm_omni_rollout_adapter.py   # BooguImageDiffusionNFTPipeline
```

Common utilities (`boogu_timestep_from_scheduler`, `apply_boogu_text_cfg`, `get_boogu_freqs_cis`, `build_boogu_native_scheduler`) will be imported directly from `verl_omni.pipelines.boogu_image_flow_grpo.common` to maintain strict code reuse.

### 4.2 Rollout Adapter (`vllm_omni_rollout_adapter.py`)
```python
@VllmOmniPipelineBase.register("BooguImagePipeline", algorithm="diffusion_nft")
class BooguImageDiffusionNFTPipeline(QwenImageTokenIdPromptMixin, BooguImagePipeline):
    """Rollout pipeline for Boogu-Image (T2I & Edit) used by DiffusionNFT."""

    diffusion_io_spec = DiffusionIOSpec(primary=MediaSpec("image"))
    supports_request_batch = True

    def forward(self, req, ...):
        # 1. Unpack prompt IDs and optional reference images
        # 2. Encode prompts via Qwen3-VL MLLM; build ref_image_hidden_states if Edit
        # 3. Run standard deterministic ODE / Euler diffuse loop (no SDE log-probs)
        # 4. Decode final latents to PIL image for reward scoring
        # 5. Return rollout_output with clean latents:
        return rollout_output(
            media=decoded_images,
            prompt_embeddings={
                "prompt_embeds": prompt_embeds,
                "prompt_embeds_mask": prompt_embeds_mask,
                "negative_prompt_embeds": negative_prompt_embeds,
                "negative_prompt_embeds_mask": negative_prompt_embeds_mask,
            },
            rl={
                "latents_clean": latents.float(),
                "train_timesteps": timesteps.unsqueeze(0).expand(latents.shape[0], -1),
                **({"condition_image_latents": condition_image_latents} if condition_image_latents is not None else {}),
            },
            to_cpu=True,
        )
```

### 4.3 Training Adapter (`diffusers_training_adapter.py`)
```python
@DiffusionModelBase.register("BooguImagePipeline", algorithm="diffusion_nft")
class BooguImageDiffusionNFT(DiffusionModelBase):
    """Forward-process Boogu-Image adapter used by DiffusionNFT."""

    @classmethod
    def prepare_model_inputs(cls, module, model_config, latents, timesteps, prompt_embeds, prompt_embeds_mask, ...):
        # 1. Map timesteps to Boogu space: t_boogu = 1 - sigma
        # 2. Fetch cached RoPE tables: get_boogu_freqs_cis(...)
        # 3. Format reference condition latents if present in micro_batch
        # 4. Construct instruction_hidden_states and instruction_attention_mask dicts
        ...

    @classmethod
    def forward(cls, module, model_config, model_inputs, negative_model_inputs=None):
        # 1. Forward module with positive inputs
        pred = module(**model_inputs)[0]
        # 2. Apply sequential text CFG if guidance_scale > 1.0
        if model_config.pipeline.guidance_scale > 1.0 and negative_model_inputs is not None:
            neg_pred = module(**negative_model_inputs)[0]
            pred = apply_boogu_text_cfg(pred, neg_pred, model_config.pipeline.guidance_scale)
        # 3. Negate velocity so pred conforms to diffusers target (epsilon - x0)
        return pred.neg()
```

### 4.4 Module Export & Auto-Registration
Add star-imports in `verl_omni/pipelines/__init__.py`:
```python
from . import boogu_image_diffusion_nft
from .boogu_image_diffusion_nft import *
__all__ += list(boogu_image_diffusion_nft.__all__)
```

---

### 5.1 CPU Unit Tests (`tests/pipelines/test_boogu_image_diffusion_nft_on_cpu.py`)
- **Registry Test**: Assert `@DiffusionModelBase.get_class("BooguImagePipeline", "diffusion_nft")` and `@VllmOmniPipelineBase.get_class("BooguImagePipeline", "diffusion_nft")` resolve to the new adapters.
- **Velocity Inversion Contract**: Validate that `forward()` returns $-v_{\text{boogu}}$ and that `x0_pred = xt - t * pred` matches $x_0$ within floating point tolerance.
- **T2I & Edit Input Preparation**: Verify correct shaping for both unconditional/T2I requests and Edit requests with `condition_image_latents`.
- **CFG Consistency**: Ensure text CFG matches the upstream formula without NaN/drift.

### 5.2 E2E Smoke Tests
- Reuse `tests/special_e2e/build_boogu_image_tiny_random.py` to create a lightweight synthetic checkpoint.
- Add `tests/special_e2e/run_diffusionnft_boogu_image.sh` covering both `MODE=t2i` and `MODE=edit`.
- Wire into `tests/gpu_smoke/run_gpu_smoke_diffusion_e2e.sh`.

### 5.3 Recipes & Examples
- Provide reference training scripts:
  - `examples/diffusionnft_trainer/boogu_image/run_boogu_image_ocr_lora.sh`
  - `examples/diffusionnft_trainer/boogu_image/run_boogu_image_edit_ocr_lora.sh`
- Reuse existing data conversion scripts (`examples/flowgrpo_trainer/data_process/boogu_image_ocr.py` & `boogu_image_edit_ocr.py`).

---

## 6. Phased Rollout Plan

- [ ] **Phase 1: Adapter Implementation & CPU Contracts**
  - Create `verl_omni/pipelines/boogu_image_diffusion_nft/`.
  - Add CPU unit tests in `tests/pipelines/test_boogu_image_diffusion_nft_on_cpu.py`.
- [ ] **Phase 2: E2E validation & Example Recipes & Documentation**
  - Add example scripts under `examples/diffusionnft_trainer/boogu_image/` after E2E training to validate the convergence and performance
  - Update `docs/start/models.md` and `README.md` to reflect DiffusionNFT support for Boogu-Image.

---

## 7. Open Questions for Discussion

1. **Old-Policy LoRA EMA Schedule**: For Qwen-Image, `delayed_linear_to_0_999` and `copy` schedules are supported. Should Boogu-Image default to `copy` or an EMA decay schedule?
2. **Text Guidance in Trainer Forward Pass**: Upstream inference uses text guidance scale 4.0. Should the actor training pass evaluate unconditional negative prompts (doubling DiT forward passes) or train predominantly on unconditional guidance (`guidance_scale=1.0`) as is common in DiffusionNFT literature?

**Comment by OnePunchMonk** [comment]
I'd like to claim §4.3 (training adapter: velocity inversion, timestep mapping, text CFG) only, for now.

Holding off on 4.1, 4.2, and 4.4: 4.2's rollout adapter sketch is written against the `rollout_output(media=..., prompt_embeddings={...}, rl={...})` shape, which #480 is replacing with a typed `OmniRolloutRequest`, and #556/#557 are further [BREAKING] changes to prompt wire fields and media artifact paths on top of that. Building the rollout adapter now means rewriting it once that series lands.

**Comment by SamitHuang** [comment]
I think holding 4.2 until #480 + #556/#557 is optional, not required. Implementing 4.3 training adapter alone can't verify E2E training. I'd suggest to work on E2E training for reward converge at first without being blocked by the WIP diffusion rollout refactor. Then migrate 4.2 in a follow-up after #480 / #556 / #557 merge.  @OnePunchMonk

**Comment by SamitHuang** [comment]
@OnePunchMonk Are you going to work on E2E training?

**Comment by OnePunchMonk** [comment]
sure!

**Comment by OnePunchMonk** [comment]
@SamitHuang pls review #568
